来源：[kate的CSDN](https://blog.csdn.net/Guo_ya_nan/article/details/80214496)

# 一、项目背景
糖尿病（尤其是2型糖尿病）算得上是21世纪最大的流行病，在我国更是**情况堪忧**。20世纪20年代之前，人们闻“糖”色变，因为无法治疗，患病就相当于慢性死亡。这种病会让血糖升高，伴随的症状有异常口渴、尿频、容易疲劳，严重的并发症有中风、失明、截肢、肾衰竭、甚至心肌梗塞。好在1920s，一种能降低血糖的胰腺分泌物 胰岛素（insulin）被 Frederick Banting 发现，他也因此获得1923年的诺贝尔医学和生理学奖。 
## 1.1胰岛素工作原理
我们吃下去的食物大部分转化为葡萄糖（glucose），为身体提供能量。位于胃（stomach）附近的胰腺（pancreas），会产生一种叫做胰岛素（insulin）的激素（hormone），这种激素能促进葡萄糖被人体细胞吸收。如果一个得了糖尿病，就意味着无法分泌足够的胰岛素（insulin），或者胰岛素不能发挥正常功能，血糖便由此增多。 

![png](./image/yidaosu.png)

虽然胰岛素不能治愈糖尿病，但仍然是医学史上最大的重大发现之一，挽救了很多人的生命。不过当时胰岛素默认是针管注射的，一天要注射多次，现在依然如此；但这不方便也不舒服，如果能口服胰岛素就好了，比如制成药片或胶囊，随身携带多方便。

所以，**口服胰岛素**(oral insulin)一直以来都是活跃的研究领域。然而，最大的拦路石在于胰岛素本质上是一种蛋白质，无法通过厚厚的胃黏膜。

## 1.2临床试验及其指标
在这个背景下，我们将接触一份 II 期临床试验数据，涉及一种新型口服胰岛素叫做 __Auralin__。Auralin 的研究员认为他们研发的胶囊能解决胃黏膜问题。

II 期试验会对药品的功效和剂量反应进行测试，还会找出药品常见的短期副作用（也即不良反应），通常涉及数百位患者。 
![png](./image/1.2.png)
在该试验中，我们有350名患者，分为两组，进行双盲试验[^footnote]。其中175人服用 Auralin 进行治疗，另外一半人则用市面上流行的注射型胰岛素 Novodra 进行治疗。通过比较两种药物的关键指标（key metrics），可以判断 Auralin 是否有效。最重要的指标就是 HbA1c 水平。HbA1c 是血液中的一种物质（糖化血红蛋白），能够衡量过去几个月来人体血糖水平的控制情况，正常值为4%-6%。如果 Auralin 能够让 HbA1c 水平降到跟注射型胰岛素 Novodra 相似的程度，就能表明 Auralin 是有效的，能够提高患者的生活质量。(比如：能引起0.4%的下降就可以认为是成功的)

![png](./image/1.22.png)
> 对数据集的几点说明： 
  - Auralin 和 Novodra 都不是真的胰岛素产品，是模拟数据； 
  - 为了合理性，数据集是在咨询专业医生的情况下建立的； 
  - 由官方授意，密切模拟了真正的[新型吸入胰岛素 Afrezza](http://care.diabetesjournals.org/content/38/12/2266.long)的临床试验； 
  - 模拟了[医疗保健数据中的真实常见数据质量问题](https://www.justassociates.com/application/files/2514/9133/9342/TheProblemwithHealthcare_2015.pdf)； 
  - 患者信息是使用[虚假姓名生成器](https://www.fakenamegenerator.com/order.php)生成。


## 1.3医疗数据 常见问题及原因
医疗数据中经常存在问题：数据冗余、数据缺失、数据不准确等。在我们的临床试验数据也存在以上问题。

![png](./image/1.3.png)

# 二、数据评估(Assess)
Unclean data 存在两类问题：数据质量低，数据不整洁,分别从这两方面分别进行评估。
## 2.1 质量评估(Quality)

为了得到尽可能多的练习，这里先用目测的方法进行数据评估，再用编程方法进行数据评估。

### 2.1.1目测评估

导入数据：

In [1]:
import pandas as pd

patients = pd.read_csv('./data/patients.csv')
treatments = pd.read_csv('./data/treatments.csv')
adverse_reactions = pd.read_csv('./data/adverse_reactions.csv')

查看数据:

In [2]:
#display the patinets table
patients.head(3)

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   

                                 contact  birthdate  weight  height   bmi  
0   951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1  PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2       402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8

![png](./image/2.1patients.png)

In [3]:
treatments.head(3)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN

![png](./image/2.1treatments.png)

In [4]:
adverse_reactions.head(3)

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia

![png](./image/2.1adverse.png)

问题记录：  

质量问题  
**patients表**  
- zip code为浮点数，应该修改为字符串。  
- zip code有时有四位数。  
- tim neudorf高度为72，应修改为27英寸  
- state有时为全程，有时为缩写  
- dsvid gustafsson拼写错误  

**treatments表**  
- 部分缺失HbA1c值  
- staring和ending中字母“u”混乱  
- given names和surnames都应变为小写    
- 缺失记录，（350代替为280）  

**adverse reactions**   
- given names和surnames都应变为小写    

### 2.1.2编程评估
编程评估要用到代码，而不是对整个数据进行浏览。就` pandas `而言，它的很多函数和方法都可以用来检查数据的质量和整洁度，比如 `.sample()、.head()、.tail()、.info()、.describe()、.value_counts() `等。

对 patients 进行数据评估： 

In [5]:
patients.head(2)

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   

                                 contact  birthdate  weight  height   bmi  
0   951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1  PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2

In [6]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [7]:
patients.isnull().sum()

patient_id       0
assigned_sex     0
given_name       0
surname          0
address         12
city            12
state           12
zip_code        12
country         12
contact         12
birthdate        0
weight           0
height           0
bmi              0
dtype: int64

In [8]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [9]:
patients.sample(5)

patient_id assigned_sex given_name  surname              address  \
147         148         male   Kristian  Vukelić    4411 Libby Street   
3             4         male       Liêm     Phan  2335 Webster Street   
101         102         male       Oles  Zhdanov  1731 Chandler Drive   
203         204       female        Mùi    Lương   1778 Rodney Street   
32           33       female        Một     Phạm   2690 Pin Oak Drive   

                 city       state  zip_code        country  \
147  West Los Angeles  California   90025.0  United States   
3          Woodbridge          NJ    7095.0  United States   
101          Fordland          MO   65652.0  United States   
203         Harvester          MO   63301.0  United States   
32         Long Beach          CA   90815.0  United States   

                                        contact   birthdate  weight  height  \
147     310-268-6864KristianVukelic@teleworm.us  11/30/1958   191.4      66   
3    PhanBaLiem@jourrapide.com+1 (732) 636-8246   7/26/1951   220.9      70   
101         OlesZhdanov@fleckens.hu417-738-5084  10/25/1945   226.2      72   
203         636-442-6946LuongHongMui@einrot.com   2/29/1956   192.7      60   
32        PhamThiBichMot@dayrep.com562-985-4582   1/30/1997   147.6      69   

      bmi  
147  30.9  
3    31.7  
101  30.7  
203  37.6  
32   21.8

In [10]:
print(type(patients['birthdate'][0]),type(patients['state'][0]))

<class 'str'> <class 'str'>


有五个列缺失值均为12，经过查看，缺失值位于12个记录中，其他列无缺失值。

#### 问题记录：

**patients table**   
丢失的人口统计学数据（address，contact列）共有12个记录  
错误的数据类型（assigned_sex,state,zip_code,birthdate列）  
phone number位置错误

**`treatments`进行数据评估:**

In [11]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [12]:
treatments.sample(3)

given_name   surname    auralin    novodra  hba1c_start  hba1c_end  \
131      søren  sørensen          -  41u - 39u         7.93       7.59   
214      erick     pinto  44u - 54u          -         7.88       7.49   
101     markus   solberg          -  35u - 30u         7.95       7.57   

     hba1c_change  
131          0.34  
214          0.39  
101          0.38

In [13]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

#### 问题记录
treatments table  
- 数据类型错误（auralin	novodra）     
- 含有缺失值，108个（hba1c_change）

In [14]:
patients.surname.value_counts().head()

Doe         6
Jakobsen    3
Taylor      3
Dratchev    2
Tucker      2
Name: surname, dtype: int64

In [15]:
patients.address.value_counts().head()

123 Main Street          6
2778 North Avenue        2
2476 Fulton Street       2
648 Old Dear Lane        2
3084 Blue Spruce Lane    1
Name: address, dtype: int64

In [16]:
patients[patients['surname'] == 'Taylor']

patient_id assigned_sex given_name surname             address      city  \
131         132       female     Sandra  Taylor  2476 Fulton Street  Rainelle   
282         283       female      Sandy  Taylor  2476 Fulton Street  Rainelle   
426         427         male    Rogelio  Taylor  4064 Marigold Lane     Miami   

    state  zip_code        country                                 contact  \
131    WV   25962.0  United States    304-438-2648SandraCTaylor@dayrep.com   
282    WV   25962.0  United States    304-438-2648SandraCTaylor@dayrep.com   
426    FL   33179.0  United States  305-434-6299RogelioJTaylor@teleworm.us   

      birthdate  weight  height   bmi  
131  10/23/1960   206.1      64  35.4  
282  10/23/1960   206.1      64  35.4  
426    9/2/1992   186.6      69  27.6

In [17]:
patients[patients['address'] == '2778 North Avenue']

patient_id assigned_sex given_name  surname            address  city  \
97           98         male    Patrick  Gersten  2778 North Avenue  Burr   
502         503         male        Pat  Gersten  2778 North Avenue  Burr   

        state  zip_code        country                               contact  \
97         NE   68324.0  United States  PatrickGersten@rhyta.com402-848-4923   
502  Nebraska   68324.0  United States  PatrickGersten@rhyta.com402-848-4923   

    birthdate  weight  height   bmi  
97   5/3/1954   138.2      71  19.3  
502  5/3/1954   138.2      71  19.3

In [18]:
patients[patients['address'].duplicated()].sort_values(by='address')

patient_id assigned_sex given_name     surname             address  \
229         230         male       John         Doe     123 Main Street   
237         238         male       John         Doe     123 Main Street   
244         245         male       John         Doe     123 Main Street   
251         252         male       John         Doe     123 Main Street   
277         278         male       John         Doe     123 Main Street   
282         283       female      Sandy      Taylor  2476 Fulton Street   
502         503         male        Pat     Gersten   2778 North Avenue   
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
242         243         male       John     O'Brian                 NaN   
249         250         male   Benjamin      Mehler                 NaN   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
278         279       female   Generosa       Cabán                 NaN   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   

            city     state  zip_code        country  \
229     New York        NY   12345.0  United States   
237     New York        NY   12345.0  United States   
244     New York        NY   12345.0  United States   
251     New York        NY   12345.0  United States   
277     New York        NY   12345.0  United States   
282     Rainelle        WV   25962.0  United States   
502         Burr  Nebraska   68324.0  United States   
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
242          NaN       NaN       NaN            NaN   
249          NaN       NaN       NaN            NaN   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
278          NaN       NaN       NaN            NaN   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   

                                        contact   birthdate  weight  height  \
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
502        PatrickGersten@rhyta.com402-848-4923    5/3/1954   138.2      71   
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
242                                         NaN   2/25/1957   205.3      74   
249                                         NaN  10/30/1951   146.5      69   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
278                                         NaN  12/16/1962 

#### 编程评估：  
经查address重复的除了id不同，其他数据值均相同，因此进行删除。  
以上姓氏中‘Doe’有6个，地址中‘123 main street’也有6个，很可能是重复，用 .duplicated() 方法筛选重复地址，发现关于‘Doe’的5条数据除 patient_id 外，其他信息都一样，现实中不可能存在这样的情况。这是一个**有效性问题**（validity issue）。

In [19]:
patients.weight.sort_values().head(3),patients.weight.sort_values().tail(3)

(210     48.8
 459    102.1
 335    102.7
 Name: weight, dtype: float64, 283    245.5
 118    254.5
 485    255.9
 Name: weight, dtype: float64)

一般人的体重不太可能是48.8磅（1 lb = 0.4536 kg），所以该数据不准确，甚至是无效的，即存在有效性问题；但是如果深入思考的话，会发现情况并非如此，其实是**一致性问题**。48.8的单位是公斤，而不是磅。下面按照48.8kg来计算其 bmi（身体质量指数），验证一下： 

In [20]:
patients[patients['weight'] == 48.8]

patient_id assigned_sex given_name   surname              address  \
210         211       female    Camilla  Zaitseva  4689 Briarhill Lane   

        city state  zip_code        country  \
210  Wooster    OH   44691.0  United States   

                                       contact   birthdate  weight  height  \
210  330-202-2145CamillaZaitseva@superrito.com  11/26/1938    48.8      63   

      bmi  
210  19.1

In [21]:
print('check:',48.8/(63*2.54/100)**2,',in table:',19.1)

check: 19.057735268405782 ,in table: 19.1


四舍五入后与记录中的 bmi 值一样，所以这条记录的体重单位有误，应该 kg。这属于**一致性问题**（consistency issue）。

In [22]:
sum(treatments.auralin.isnull()),sum(treatments.novodra.isnull())

(0, 0)

In [23]:
treatments.sample(3)

given_name   surname    auralin    novodra  hba1c_start  hba1c_end  \
60   onyekachukwu    obinna  37u - 46u          -         7.58       7.12   
16        camilla  zaitseva  28u - 37u          -         7.53       7.13   
110         suoma  koivunen          -  26u - 25u         7.59       7.13   

     hba1c_change  
60            NaN  
16            NaN  
110           NaN

计算auralin和novodra的空值个数，结果均为0，但实际上存在空值；这是一种常见的问题，将空值存储为破折号、斜杠、‘none’等，而不存储为真正的空值。这是**一个有效性问题**（validity issue），因为空值应该为空值，而不是文本。

#### 问题记录：
treatments table  
- auralin和novodra存在空值，并且以“-”存在  
- 部分错误的hba1c_change值

## 2.2 整洁度评估（Tidness）
何为整洁数据： 
 
>1. 每个变量构成一列 
2. 每项观察构成一行 
3. 每种类型的观察单元构成一个表格  
by Hadley Wickham 

#### 按照以上定义对数据整洁度进行评估：

patients 表中 contact 列的问题在于，一个列中存储了两个变量：电话号码，email地址。并记录问题。 

In [24]:
patients.head(2)

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   

                                 contact  birthdate  weight  height   bmi  
0   951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1  PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2

treatments 表的问题在于，auralin 列和 novodra 列的名称其实是变量，不应用作列标签，同时初始剂量、结束剂量应该为两个变量。并记录问题。

In [25]:
treatments.head(2)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   

   hba1c_change  
0           NaN  
1          0.97

该医学试验的观察单元应该为2个，而非3个。病人信息为一张表，药物用量及不良反应为一张表。所以，patients表保持不变，treatments表和adverse_reactions表合二为一。并记录问题。    

因为要将三个观察单位变为两个，可以查看下三张表重复的列有哪些： 

In [26]:
all_columns = pd.Series(list(patients)+list(treatments)+list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

# 三、数据清理
清理过程中，最好优先处理缺失数据的问题，即完整性问题。通过拼接（concatenate）、合并（join）、填充（impute）等手段，将数据补充完整，然后再清理其他问题。

该项目有三处缺失问题，`patients`表中有人口统计信息缺失的问题，暂时无法补充完整，先不管。`treatments`表中有两处缺失，一个是hba1c变化水平，该值可以计算获得；一个是缺少记录，可以在文件`treatments_cut.csv`中找到缺失数据。 


问题记录：

**patients表**

zip code为浮点数，应该修改为字符串。  
zip code有时有四位数。  
tim neudorf高度为72，应修改为27英寸  
state有时为全程，有时为缩写  
dsvid gustafsson拼写错误  

丢失的人口统计学数据（address，contact列）共有12个记录  
错误的数据类型（assigned_sex,state,zip_code,birthdate列）  
phone number位置错误  
john doe数据缺失  
Jakobsen Taylor Lâm Kadyrov 数据重复  
Camilla	Zaitseva重量单位实际为kg，应换算为镑   
 

**treatments表**

部分缺失HbA1c值  
staring和ending中字母“u”混乱  
given names和surnames都应变为小写  
缺失记录，（350代替为280）  
auralin和novodra存在空值，并且以“-”存在  
部分错误的hba1c_change值   

**adverse reactions**

given names和surnames都应变为小写

## 3.1 数据缺失问题(Missing Data)
先处理这两处数据缺失问题。

清理前先拷贝数据：

In [27]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

In [28]:
#code
treatments_cut = pd.read_csv('./data/treatments.csv')
treatment_clean = pd.concat([treatments_clean, treatments_cut], ignore_index=True)

In [29]:
#test
len(treatments_clean), treatments_clean.duplicated().sum()

(280, 1)

In [30]:
# 应为350

In [31]:
treatments_cut.head(3)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN

In [50]:
treatments_cut.tail(3)

given_name       surname    auralin novodra  hba1c_start  hba1c_end  \
277     mathea       lillebø  23u - 36u       -         9.04       8.67   
278     vallie        prince  31u - 38u       -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u       -         8.00       7.64   

     hba1c_change  
277          0.37  
278          0.36  
279          0.36

![](./image/treatment2.png)

In [ ]:
#code
treatments_clean['hba1c_change'] = (treatments_clean['hba1c_start'] - 
                                   treatments_clean['hba1c_end'])

In [ ]:
#test
treatments_clean['hba1c_change'].describe()

In [ ]:
treatments_clean['hba1c_change'].isnull().sum()

## 3.2 数据整洁度问题
缺失问题处理完之后，进行整洁度处理。整洁的数据可以使后续清理工作更容易进行。整洁的低质量数据 比 不整洁的低质量数据 要容易清理，所以，先解决 结构问题 再解决 质量问题。

## 3.3 数据质量问题
# 四、后记
我本人之前对于糖尿病也是懵懂无知的状态，因为做这个项目的缘故，我主动了解糖尿病，并查阅了一些患病数据；也因为有家人得2型糖尿病，更促使我关注2型糖尿病的诱因和防控手段。

其来势汹汹（甚至是爆炸性的增长）主要跟不良的生活方式有关，在此郑重提醒 青年们 养成好的生活习惯：管住嘴，迈开腿。

## 注脚
[1] 果壳网[关于中国糖尿病患者的报告](https://www.guokr.com/article/441983/)，另外还有中国疾病控制预防中心对糖尿病的一些实况报道。  
[2] 全球糖尿病社区（the global diabetes community）[关于口服胰岛素（oral insulin）的一篇文章 ](https://www.diabetes.co.uk/oral-insulin.html)

In [ ]:
patients.dropna(how='any',inplace = True)

In [ ]:
patients[patients['address'].isnull()]

'address','city', 'state', 'zip_code', 'country', 'contact'缺失均为12，其他列无缺失，检查发现此6列缺失12个值位于相同行。进行删除。

In [ ]:
patients['patient_id'].unique().shape[0],patients.shape[0]

In [ ]:
patients.sample(5)

In [ ]:
patients.describe()

In [ ]:
patients['assigned_sex'].value_counts()

In [ ]:
int(2.1)

In [ ]:
da1 = pd.DataFrame({'A':[1.0,2,3.0]})
da1

In [ ]:
da1.dtypes

In [ ]:
da1['A'] = da1['A'].astype(int)
da1

In [ ]:
da1.dtypes

In [ ]:
treatments.sample(3)